pkg/localstorev2/internal/upload/uploadstore.go

```diff
@@ -147,104 +96,320 @@ func (i pushItem) String() string {
 	}
 
-	if err := gd.storage.ChunkStore().Delete(ctx, address); err != nil {
-		return fmt.Errorf("chunk store delete chunk %q call failed: %w", address, err)
+	ui.Uploaded = now().Unix()
+	ui.TagID = u.tagID
+
+	if err := u.s.Store().Put(ui); err != nil {
+		return fmt.Errorf("store put item %q call failed: %w", ui, err)
 	}
 
-	tai := &tagIDAddressItem{
-		TagID:   gd.tagID,
-		Address: address,
+	pi := &pushItem{
+		Timestamp: ui.Uploaded,
+		Address:   chunk.Address(),
+		BatchID:   chunk.Stamp().BatchID(),
+		TagID:     u.tagID,
 	}
-	if err := gd.storage.Store().Delete(tai); err != nil {
-		return fmt.Errorf("store delete item %q call failed: %w", tai, err)
+	if err := u.s.Store().Put(pi); err != nil {
+		return fmt.Errorf("store put item %q call failed: %w", pi, err)
 	}
+
 	return nil
 }
 
-// has checks if the given address has the tagID of this getterDeleter.
-func (gd *getterDeleter) has(address swarm.Address) error {
-	tai := &tagIDAddressItem{
-		Address: address,
-		TagID:   gd.tagID,
+// Close provides the CloseWithReference interface where the session can be associated
+// with a swarm reference. This can be useful while keeping track of uploads through
+// the tags. It will update the tag. This will be filled with the Split and Seen count
+// by the Putter.
+func (u *uploadPutter) Close(addr swarm.Address) error {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	ti := &tagItem{TagID: u.tagID}
+	err := u.s.Store().Get(ti)
+	if err != nil {
+		return fmt.Errorf("failed reading tag while closing: %w", err)
 	}
-	switch exists, err := gd.storage.Store().Has(tai); {
-	case err != nil:
-		return fmt.Errorf("store has item %q call failed: %w", tai, err)
-	case !exists:
-		return storage.ErrNotFound
+
+	ti.Split = u.split
+	ti.Seen = u.seen
+
+	if !addr.IsZero() {
+		ti.Address = addr.Clone()
 	}
+
+	err = u.s.Store().Put(ti)
+	if err != nil {
+		return fmt.Errorf("failed storing tag: %w", err)
+	}
+
 	return nil
 }
 
-// ChunkGetterDeleter returns a storage.ChunkGetterDeleter
-// that restricts its operation to the given tagID.
-func ChunkGetterDeleter(s internal.Storage, tagID uint64) storage.ChunkGetterDeleter {
-	return &getterDeleter{storage: s, tagID: tagID}
+type pushReporter struct {
+	s internal.Storage
 }
 
-// ChunkPutter returns a storage.Putter which will store the given chunk.
-func ChunkPutter(s internal.Storage, tagID uint64) (storage.Putter, error) {
-	return storage.PutterFunc(func(ctx context.Context, chunk swarm.Chunk) error {
-		tai := &tagIDAddressItem{
-			Address: chunk.Address(),
-			TagID:   tagID,
-		}
-		switch exists, err := s.Store().Has(tai); {
-		case err != nil:
-			return fmt.Errorf("store has item %q call failed: %w", tai, err)
-		case exists:
-			return nil
-		}
-		if err := s.Store().Put(tai); err != nil {
-			return fmt.Errorf("store put item %q call failed: %w", tai, err)
-		}
+func NewPushReporter(s internal.Storage) storage.PushReporter {
```

Comment:
A description would be nice to have. The same applies to the `Report` method below.